README.md
# apt.postgresql.org

The main idea of this project is to provide PostgreSQL and its Extensions
packages for all supported PostgreSQL versions and debian version. That
means things like PostgreSQL 9.1 for `squeeze`, with all known debian
extensions, and for `sid` and `wheezy` too. And same for `8.3` and `8.4` and
`9.0` and `9.1` and `9.2`. Targeting `i386` and `amd64`.

https://wiki.postgresql.org/wiki/Apt

## Build System

The build system is currently based on jenkins and some scripting. `Jenkins`
is only about launching the build jobs, the real work is done in the scripts.

## Adding a new PostgreSQL version to the repository

* Update apt.postgresql.org/repo/conf/distributions.sh and re-generate the distributions file
* Update `PG_DEVEL_VERSION` in pgapt.conf
* Possibly update the Jenkins pipeline jobs to include the new version
* Update the repository master host: cd /srv/apt/apt.postgresql.org; git pull
* Create new Jenkins jobs and build them

## Changing the default PostgreSQL version

* Update `PG_MAIN_VERSION` in pgapt.conf
* Update the repository master host: cd /srv/apt/apt.postgresql.org; git pull
* Rebuild the old default version so its libpq gets added to the extra component (as it was in main before)
* Rebuild the new default version so its libpq gets added to main (as it was in the extra component before)
* Possibly update the Jenkins pipeline jobs to stop triggering beta jobs
* In postgresql-common, update debian/supported-versions
* To remove obsolete old lib packages, use
  ```
  . /srv/apt/apt.postgresql.org/pgapt.conf
  COMPONENT=11
  PKGS="libpq5 libecpg6 libecpg-compat3 libecpg-dev libpgtypes3 libpq-dev"
  for dist in $PG_SUPPORTED_DISTS; do for pkg in $PKGS; do sudo -u aptuser reprepro -C $COMPONENT remove $dist-pgdg-testing $pkg $pkg-dbgsym; done; done
  ```
* Removing obsolete PG component:
  ```
  . /srv/apt/apt.postgresql.org/pgapt.conf
  COMPONENT=11
  PKGS="libecpg6 libecpg-compat3 libecpg-dev libpgtypes3 libpq5 libpq-dev postgresql-$COMPONENT postgresql-$COMPONENT-dbg postgresql-client-$COMPONENT postgresql-contrib-$COMPONENT postgresql-doc-$COMPONENT postgresql-plperl-$COMPONENT postgresql-plpython3-$COMPONENT postgresql-plpython-$COMPONENT postgresql-pltcl-$COMPONENT postgresql-server-dev-$COMPONENT"
  for dist in $PG_SUPPORTED_DISTS; do for pkg in $PKGS; do sudo -u aptuser reprepro -C $COMPONENT remove $dist-pgdg-testing $pkg $pkg-dbgsym; done; done
  ```

## Changing the set of supported PostgreSQL versions

* In postgresql-common, update debian/supported-versions
* Trigger a postgresql-common build
* Upgrade postgresql-common on pgdgbuild.dus.dg-i.net because
  generate-pgdg-source uses the list when generating debian/control from
  debian/control.in
* Restrict deprecated version to `dist-filter: '(distribution=="sid")'`
* Announce change on wiki and the mailing list

## Adding a new distribution

Update these files:

* pgapt.conf
* repo/conf/distributions.sh and generate repo/conf/distributions from it
* repo/conf/incoming
* jenkins/pgapt-jobs.yaml

Update repository master host:

* cd /srv/apt/apt.postgresql.org; git pull
* Seed initial packages: reprepro copysrc $newdist-pgdg-testing $olddist-pgdg-testing postgresql-common pgdg-keyring pgdg-buildenv; postgresql-x.y from sid-pgdg-testing (or whatever seems appropriate)
* possibly: reprepro copymatched $newdist-pgdg $newdist-pgdg-testing \*
* or: reprepro export $newdist-pgdg (so apt can run update on it even if it's still empty)
* Wait until the mirror sync has run (xx:17)

Update Jenkins hosts:

* Run the "apt.postgresql.org" job
* cd jenkins/ansible; ./setup-buildd.yml
* Run the "sbuild-update" job

Do the paperwork:

* Update wiki pages
* Update pgdg/apt.postgresql.org.sh in postgresql-common
* Send announcement


debian/changelog
pgdg-buildenv (29) UNRELEASED; urgency=medium

  * Move "Destroy all schroot sessions on reboot" to ansible.

 -- Christoph Berg <myon@debian.org>  Wed, 30 Oct 2019 16:42:50 +0100

pgdg-buildenv (28) sid-pgdg; urgency=medium

  * Destroy all schroot sessions on reboot.

 -- Christoph Berg <myon@debian.org>  Wed, 30 Oct 2019 16:24:47 +0100

pgdg-buildenv (27) sid-pgdg; urgency=medium

  * Move llvm/clang dependency from pgdg-buildenv to sbuild-update.
  * Remove sudoers.d override obsolete since debhelper 8.
    Thanks to nthykier for pointing this out!

 -- Christoph Berg <myon@debian.org>  Thu, 18 Jul 2019 21:28:19 +0200

pgdg-buildenv (26) sid-pgdg; urgency=medium

  * Install lintian in chroots, sbuild installs it anyway.
  * Install clang and llvm-dev.
  * Check that pgdg-buildenv is installable.

 -- Christoph Berg <myon@debian.org>  Thu, 18 Jul 2019 16:09:02 +0200

pgdg-buildenv (25) sid-pgdg; urgency=medium

  * Always use our postgresql-common packages.

 -- Christoph Berg <christoph.berg@credativ.de>  Mon, 15 Oct 2018 16:36:15 +0200

pgdg-buildenv (24) sid-pgdg; urgency=medium

  * adt-sbuild: Set LC_ALL=C.UTF-8 in old adt-run interface.

 -- Christoph Berg <christoph.berg@credativ.de>  Mon, 25 Jun 2018 10:55:19 +0200

pgdg-buildenv (23) sid-pgdg; urgency=medium

  * Install pkg-perl-autopkgtest.

 -- Christoph Berg <christoph.berg@credativ.de>  Wed, 30 May 2018 15:18:03 +0200

pgdg-buildenv (22) sid-pgdg; urgency=medium

  * Install autodep8, and run autopkgtest unconditionally.

 -- Christoph Berg <christoph.berg@credativ.de>  Fri, 01 Dec 2017 13:42:34 +0100

pgdg-buildenv (21) sid-pgdg; urgency=medium

  * Update adt-sbuild to support /usr/bin/autopkgtest.

 -- Christoph Berg <christoph.berg@credativ.de>  Tue, 19 Sep 2017 11:22:40 +0200

pgdg-buildenv (20) sid-pgdg; urgency=medium

  * Update sbuild-pgdg-components to use sudo only when necessary.

 -- Christoph Berg <myon@debian.org>  Wed, 06 Sep 2017 18:14:57 +0200

pgdg-buildenv (19) sid-pgdg; urgency=medium

  * Add libfile-fcntllock-perl.

 -- Christoph Berg <myon@debian.org>  Sat, 27 Aug 2016 13:11:57 +0200

pgdg-buildenv (18) sid-pgdg; urgency=medium

  * Add aptitude.

 -- Christoph Berg <christoph.berg@credativ.de>  Mon, 11 Jan 2016 11:17:43 +0100

pgdg-buildenv (17) sid-pgdg; urgency=medium

  * Add build-essential.
  * Don't install sources.list.d/pgdg.list from here, we can't easily figure
    out the current release name from the postinst because sid has "jessie" in
    there.
  * Add /usr/bin/adt-sbuild and sbuild-pgdg-components.
  * Add sudo and /etc/sudoers.d/sbuild.
  * Add /usr/sbin/policy-rc.d.
  * createcluster.conf: Set create_main_cluster = false.
  * Merge repository into pgapt.git.

 -- Christoph Berg <myon@debian.org>  Sat, 10 Oct 2015 15:27:33 +0200

pgdg-buildenv (16) sid-pgdg; urgency=medium

  * Add /etc/dpkg/dpkg.cfg.d/01unsafeio.

 -- Christoph Berg <myon@debian.org>  Tue, 06 Jan 2015 19:55:41 +0100

pgdg-buildenv (15) sid-pgdg; urgency=medium

  * Pin the pgdg repos to 500 so we don't miss newer distribution packages.

 -- Christoph Berg <myon@debian.org>  Fri, 26 Sep 2014 17:36:16 +0200

pgdg-buildenv (14) sid-pgdg; urgency=medium

  * Fix postinst to exit cleanly.

 -- Christoph Berg <christoph.berg@credativ.de>  Wed, 19 Feb 2014 12:28:22 +0100

pgdg-buildenv (13) sid-pgdg; urgency=medium

  * Bump dh minimum version to 9.
  * Depend on dpkg-dev (>= 1.16.1~) | hardening-wrapper (lucid needs h-w).
  * Install locales for the PostgreSQL testsuite.

 -- Christoph Berg <christoph.berg@credativ.de>  Wed, 19 Feb 2014 12:14:42 +0100

pgdg-buildenv (12) sid-pgdg; urgency=medium

  * Use local approx cache on buildhost.

 -- Christoph Berg <christoph.berg@credativ.de>  Mon, 17 Feb 2014 16:56:24 +0100

pgdg-buildenv (11) sid-pgdg; urgency=medium

  * Add logrotate.

 -- Christoph Berg <myon@debian.org>  Sat, 01 Feb 2014 23:45:52 +0100

pgdg-buildenv (10) sid-pgdg; urgency=medium

  * Add newpid.

 -- Christoph Berg <christoph.berg@credativ.de>  Thu, 09 Jan 2014 11:28:17 +0100

pgdg-buildenv (9) sid-pgdg; urgency=medium

  * Update sources.list and apt preferences.

 -- Christoph Berg <myon@debian.org>  Thu, 26 Dec 2013 18:03:14 +0200

pgdg-buildenv (8) sid-pgdg; urgency=medium

  * Add autopkgtest.
  * Install /etc/postgresql-common/createcluster.conf.
  * Install /etc/apt/apt.conf.d/20norecommends.

 -- Christoph Berg <myon@debian.org>  Thu, 26 Dec 2013 16:51:07 +0200

pgdg-buildenv (7) sid-pgdg; urgency=low

  * Add netcat-openbsd.

 -- Christoph Berg <myon@debian.org>  Sun, 16 Dec 2012 15:48:54 +0100

pgdg-buildenv (6) sid-pgdg; urgency=low

  * Add eatmydata.
  * Depends: debhelper (>= 8~).

 -- Christoph Berg <myon@debian.org>  Fri, 07 Sep 2012 15:19:15 +0200

pgdg-buildenv (5) sid-pgdg; urgency=low

  * Depend on dependencies of postgresql-server-dev-all instead of the package
    itself. (We should not pull in any libpq or pg_config by default.)
  * Add locales needed by the postgresql-testsuite.

 -- Christoph Berg <myon@debian.org>  Wed, 05 Sep 2012 10:50:14 +0200

pgdg-buildenv (4) sid-pgdg; urgency=low

  * Use the correct syntax: Acquire::Languages "none";

 -- Christoph Berg <myon@debian.org>  Tue, 04 Sep 2012 21:46:53 +0200

pgdg-buildenv (3) sid-pgdg; urgency=low

  * Add Depends: debhelper postgresql-server-dev-all.

 -- Christoph Berg <myon@debian.org>  Tue, 04 Sep 2012 19:42:49 +0200

pgdg-buildenv (2) sid-pgdg; urgency=low

  * Skip downloading apt translations. Suggested by Mika Prokop.

 -- Christoph Berg <myon@debian.org>  Tue, 04 Sep 2012 19:28:50 +0200

pgdg-buildenv (1) sid-pgdg; urgency=low

  * Divert /usr/bin/mandb.

 -- Christoph Berg <myon@debian.org>  Sun, 26 Aug 2012 19:49:11 +0200


jenkins/sbuild-update.sh
#!/bin/sh

set -eu

error () {
  echo "Error: $@" >&2
  exit 1
}

# use "sbuild" chroot here (default bind-mounts /root)
chroot="source:${distribution:=sid}-pgdg-${architecture:=amd64}-sbuild"
chroot_path="/home/chroot/$distribution-$architecture"

# read pgapt config
for dir in . .. /home/jenkins/jenkins/workspace/apt.postgresql.org /home/buildd/workspace/apt.postgresql.org; do
  test -f $dir/pgapt.conf || continue
  . $dir/pgapt.conf
  break
done
set_dist_vars $distribution

# check if chroot is configured in sbuild
schroot -l | grep -q $chroot || \
  error "There is no schroot definition for $chroot (run schroot-config.sh first)"

PGDG_SH=$(mktemp /var/tmp/pgdg.XXXXXX.sh)
trap "rm -f $PGDG_SH" 0 2 3 15
cat < /usr/share/postgresql-common/pgdg/apt.postgresql.org.sh > $PGDG_SH

# build package and run autopkgtests inside the chroot
# (lock against concurrent apt-get update/upgrade operations and builds)
LOCKDIR="/var/lock/sbuild-package"
if ! test -d $LOCKDIR; then
  mkdir $LOCKDIR
  chgrp sbuild $LOCKDIR
  chmod 3775 $LOCKDIR
fi
umask 002

(
  cd /
  set -x
  flock --exclusive 9 # lock against concurrent access

  # create chroot if it doesn't exist yet
  if ! test -d $chroot_path; then
    echo "Creating chroot in $chroot_path"
    if ! test -e /usr/share/debootstrap/scripts/$distribution; then
      case $DISTRO in
        debian) sudo ln -sv sid /usr/share/debootstrap/scripts/$distribution ;;
        ubuntu) sudo ln -sv gutsy /usr/share/debootstrap/scripts/$distribution ;;
      esac
    fi
    sudo debootstrap --variant=buildd --arch=$architecture $distribution $chroot_path $mirror
  fi

  # do the update
  schroot -u root -c $chroot -- sh <<-EOF
	set -ex
	
	# configure dpkg and apt
	test -e /etc/dpkg/dpkg.cfg.d/01unsafeio || echo force-unsafe-io | tee /etc/dpkg/dpkg.cfg.d/01unsafeio
	test -e /etc/apt/apt.conf.d/20norecommends || echo 'APT::Install-Recommends "false";' | tee /etc/apt/apt.conf.d/20norecommends
	test -e /etc/apt/apt.conf.d/50i18n || echo 'Acquire::Languages { none; };' | tee /etc/apt/apt.conf.d/50i18n
	test -e /etc/apt/apt.conf.d/60releaseinfo || echo 'Acquire::AllowReleaseInfoChange "true" { Suite "true"; };' | tee /etc/apt/apt.conf.d/60releaseinfo # don't complain when testing gets released as stable
	rm -f /var/lib/apt/lists/*_Translation-*

	# run apt.postgresql.org.sh
	if ! test -f $PGDG_SH; then
		echo "$PGDG_SH not found in the chroot, did you configure a /var/tmp bindmount?"
		exit 1
	fi
	if ! test -f /etc/apt/sources.list.d/pgdg.list; then
		apt-get install -y gnupg
		chmod +x $PGDG_SH
		echo yes | $PGDG_SH
	fi

	# write sources lists
	echo "deb $mirror $distribution main" > /etc/apt/sources.list
	echo "deb-src $mirror_src $distribution main" >> /etc/apt/sources.list
	echo "deb $apt1 $distribution-pgdg main" > /etc/apt/sources.list.d/pgdg.list
	echo "deb $apt2 $distribution-pgdg-testing main" >> /etc/apt/sources.list.d/pgdg.list
	echo "deb-src $apt2 $distribution-pgdg-testing main" >> /etc/apt/sources.list.d/pgdg.list
	case $DISTRO in
	  ubuntu) # libossp-uuid-dev is in universe on vivid+
	    echo "deb $mirror $distribution universe" > /etc/apt/sources.list.d/universe.list
	    echo "deb-src $mirror_src $distribution universe" >> /etc/apt/sources.list.d/universe.list
	    echo "deb $mirror $distribution-updates main universe" > /etc/apt/sources.list.d/updates.list
	    echo "deb-src $mirror_src $distribution-updates main universe" >> /etc/apt/sources.list.d/updates.list
	    echo "deb $mirror $distribution-security main universe" > /etc/apt/sources.list.d/security.list
	    echo "deb-src $mirror_src $distribution-security main universe" >> /etc/apt/sources.list.d/security.list
	    ;;
	  *)
	    if [ "${dist_security:-}" ]; then
	      echo "deb ${security:-} ${dist_security:-} main" > /etc/apt/sources.list.d/security.list
	      echo "deb-src ${security:-} ${dist_security:-} main" >> /etc/apt/sources.list.d/security.list
	    fi
	    ;;
	esac
	if [ "$HAS_BACKPORTS" ]; then
	  echo "${mirror_backports:-}" > /etc/apt/sources.list.d/backports.list
	  if [ -d /var/lib/apt/backports ]; then
	    rm -f /var/lib/apt/lists/*backports* # clean up if last run failed
	    cp -al /var/lib/apt/backports/* /var/lib/apt/lists
	  fi
	else
	  rm -f /etc/apt/sources.list.d/backports.list*
	fi

	# tell debconf and ucf not to ask any questions
	export DEBIAN_FRONTEND=noninteractive
	export UCF_FORCE_CONFFNEW=y UCF_FORCE_CONFFMISS=y

	# try update twice because ubuntu mirrors keep acting up
	apt-get -y update || { sleep 60; apt-get -y update; }

	# save backports lists
	rm -rf /var/lib/apt/backports /etc/apt/sources.list.d/backports.list.disabled
	if [ "$HAS_BACKPORTS" ]; then
	  # install lintian from backports (necessary because
	  # /usr/share/doc/lintian/lintian.html changes filetype in
	  # stretch/buster/bullseye, and upgrading on overlayfs fails)
	  apt-get -y install -t $distribution-backports lintian

	  mv /etc/apt/sources.list.d/backports.list /etc/apt/sources.list.d/backports.list.disabled
	  mkdir -p /var/lib/apt/backports
	  mv /var/lib/apt/lists/*backports* /var/lib/apt/backports
	fi
	
	[ -x /usr/bin/eatmydata ] && eatmydata="eatmydata"
	\$eatmydata apt-get -y -o DPkg::Options::=--force-confnew dist-upgrade
	#case $distribution in
	#  squeeze) \$eatmydata apt-get -y -o DPkg::Options::=--force-confnew install debhelper/${distribution}-backports ;;
	#esac
	# install llvm/clang (not in pgdg-buildenv because it's architecture-specific)
	case $distribution in
	  jessie) ;;
	  stretch|bionic|xenial)
	    case $architecture in amd64|i386) apt-get -y install llvm-6.0-dev clang-6.0 ;; esac ;;
	  buster|disco) apt-get -y install llvm-7-dev clang-7 ;;
	  *) apt-get -y install llvm-9-dev clang-9 libllvm7-;;
	esac
	# bionic-updates has libllvm7, but we don't want that yet (2019-09-03)
	[ "$distribution" = "bionic" ] && apt-get remove --purge -y libllvm7
	\$eatmydata apt-get -y -o DPkg::Options::=--force-confnew install pgdg-buildenv pgdg-keyring
	eatmydata apt-get -y autoremove --purge
	apt-get clean
	
	# remove libreadline-dev if present
	if dpkg -l 'libreadline*-dev' | grep -q '^ii'; then
	  apt-get remove -y 'libreadline*-dev'
	fi

	dpkg -l 'libpq*' 'newpid' 'pgdg*' 'postgresql*' 'libreadline*' 'libedit*' || :

	# don't create any cluster on PostgreSQL installation
	grep -q '^create_main_cluster = false' /etc/postgresql-common/createcluster.conf || sed -i -e 's/.*create_main_cluster.*/create_main_cluster = false/' /etc/postgresql-common/createcluster.conf

	# pre-generate autopkgtest signing key
	if ! test -f /root/.cache/autopkgtest/secring.gpg &&
	  dpkg --compare-versions \$(dpkg-query -W -f '\${Version}' autopkgtest) lt 3.16; then
	  umask 077
	  mkdir -p /root/.cache/autopkgtest
	  cd /root/.cache/autopkgtest
	  cat > key-gen-params <<-EOT
		Key-Type: DSA
		Key-Length: 1024
		Key-Usage: sign
		Name-Real: autopkgtest per-run key
		Name-Comment: do not trust this key
		Name-Email: autopkgtest@example.com
		EOT
	  gpg --homedir=/root/.cache/autopkgtest --batch --no-random-seed-file --gen-key key-gen-params
	fi

	EOF
) 9> $LOCKDIR/$distribution-$architecture.lock


jenkins/schroot-config.sh
#!/bin/sh

# union-type=overlay needs Linux 4.x and schroot >= 1.6.10-2~.

# ./schroot-config.sh | sudo tee /etc/schroot/chroot.d/sbuild.conf

set -eu

# seed PG_SUPPORTED_DISTS if pgapt.conf is not yet installed
PG_SUPPORTED_DISTS="sid"

# read pgapt config
for dir in . .. /home/jenkins/jenkins/workspace/apt.postgresql.org /home/buildd/workspace/apt.postgresql.org; do
  test -f $dir/pgapt.conf || continue
  . $dir/pgapt.conf
  break
done

case $(dpkg --print-architecture) in
	amd64) ARCHS="amd64 i386" ;;
	ppc64el) ARCHS="ppc64el" ;;
esac

for dist in $PG_SUPPORTED_DISTS; do
	for arch in $ARCHS; do

		# if arguments are given, execute as command in enviroment
		# ./schroot-config.sh ./sbuild-update.sh
		if [ "$*" ]; then
			echo "### distribution=$dist architecture=$arch ###"
			export distribution=$dist architecture=$arch
			"$@"
			continue
		fi

		body="$(cat <<-EOF
			type=directory
			groups=sbuild
			root-groups=sbuild
			source-groups=sbuild
			source-root-groups=sbuild
			directory=/home/chroot/$dist-$arch
			union-type=overlay
			union-overlay-directory=/var/run
			EOF
			[ $arch = i386 ] && echo "personality=linux32"
			echo
		)"

		echo "[$dist-$arch]"
		aliases="$dist"
		[ $dist = sid ] && [ $arch = amd64 ] && aliases="$aliases,default"
		[ $arch = amd64 ] && echo "aliases=$aliases"
		echo "$body"
		echo

		echo "[$dist-pgdg-$arch-sbuild]"
		[ $dist = sid ] && echo "aliases=unstable-$arch-sbuild,experimental-$arch-sbuild"
		echo "profile=sbuild"
		echo "$body"
		echo
	done
done


repo/conf/distributions.sh
#!/bin/sh

DISTRIBUTIONS="sid bullseye buster stretch jessie wheezy squeeze lenny etch
	focal eoan disco cosmic bionic zesty xenial wily utopic trusty saucy precise lucid"
FLAVORS="pgdg pgdg-testing"

for DIST in $DISTRIBUTIONS ; do
	for FLAVOR in $FLAVORS ; do
		D="$DIST-$FLAVOR"
		ARCHS="amd64 i386"
		case $DIST in
			# Debian
			lenny|etch)
				COMPONENTS="main 8.2 8.3 8.4 9.0 9.1 9.2" ;;
			squeeze)
				COMPONENTS="main 8.2 8.3 8.4 9.0 9.1 9.2 9.3 9.4 9.5" ;;
			wheezy)
				COMPONENTS="main 8.2 8.3 8.4 9.0 9.1 9.2 9.3 9.4 9.5 9.6 10" ;;
			jessie) ARCHS="amd64 i386 ppc64el"
				COMPONENTS="main         8.4 9.0 9.1 9.2 9.3 9.4 9.5 9.6 10 11 12" ;;
			stretch) ARCHS="amd64 i386 ppc64el"
				COMPONENTS="main                     9.2 9.3 9.4 9.5 9.6 10 11 12 13" ;;
			buster) ARCHS="amd64 i386 ppc64el"
				COMPONENTS="main                         9.3 9.4 9.5 9.6 10 11 12 13" ;;
			bullseye) ARCHS="amd64 ppc64el"
				COMPONENTS="main                             9.4 9.5 9.6 10 11 12" ;;
			sid)    ARCHS="amd64 i386 ppc64el"
				COMPONENTS="main 8.2 8.3 8.4 9.0 9.1 9.2 9.3 9.4 9.5 9.6 10 11 12 13" ;;
			# Ubuntu
			lucid)
				COMPONENTS="main     8.3 8.4 9.0 9.1 9.2 9.3 9.4" ;;
			precise)
				COMPONENTS="main 8.2 8.3 8.4 9.0 9.1 9.2 9.3 9.4 9.5 9.6" ;;
			saucy)
				COMPONENTS="main 8.2 8.3 8.4 9.0 9.1 9.2 9.3 9.4" ;;
			trusty) ARCHS="amd64 i386 ppc64el"
				COMPONENTS="main         8.4 9.0 9.1 9.2 9.3 9.4 9.5 9.6 10 11" ;;
			utopic)
				COMPONENTS="main         8.4 9.0 9.1 9.2 9.3 9.4 9.5" ;;
			wily)
				COMPONENTS="main                 9.1 9.2 9.3 9.4 9.5 9.6" ;;
			xenial) ARCHS="amd64 i386 ppc64el"
				COMPONENTS="main                 9.1 9.2 9.3 9.4 9.5 9.6 10 11 12 13" ;;
			zesty) ARCHS="amd64 i386 ppc64el"
				COMPONENTS="main                     9.2 9.3 9.4 9.5 9.6 10" ;;
			bionic) ARCHS="amd64 i386 ppc64el"
				COMPONENTS="main                         9.3 9.4 9.5 9.6 10 11 12 13" ;;
			cosmic) ARCHS="amd64"
				COMPONENTS="main                         9.3 9.4 9.5 9.6 10 11" ;;
			disco)  ARCHS="amd64"
				COMPONENTS="main                             9.4 9.5 9.6 10 11 12" ;;
			eoan)   ARCHS="amd64"
				COMPONENTS="main                             9.4 9.5 9.6 10 11 12" ;;
			focal)  ARCHS="amd64 ppc64el"
				COMPONENTS="main                             9.4 9.5 9.6 10 11 12 13" ;;
			*)
				echo "$D missing in COMPONENTS list" >&2
				exit 1 ;;
		esac
		COMPONENTS=$(echo $COMPONENTS) # strip duplicate spaces
		cat <<EOF
Codename: $D
Suite: $D
Origin: apt.postgresql.org
Label: PostgreSQL for Debian/Ubuntu repository
Architectures: source $ARCHS
Components: $COMPONENTS
SignWith: ACCC4CF8
Log: $D.log
Uploaders: uploaders
DebIndices: Packages Release . .gz .bz2
UDebIndices: Packages . .gz .bz2
DscIndices: Sources Release .gz .bz2
Tracking: minimal includebuildinfos keepsources
Contents: percomponent nocompatsymlink
EOF
		case $FLAVOR in *testing)
			echo "NotAutomatic: yes"
			echo "ButAutomaticUpgrades: yes"
			;;
		esac
		echo
	done
done
